File: code/java/MyPolicy.java
import edu.yale.cs.cpsc474.GreedyThrower;
import edu.yale.cs.cpsc474.GreedyPegger;
import edu.yale.cs.cpsc474.CribbagePolicy;
import edu.yale.cs.cpsc474.CompoundPolicy;
import edu.yale.cs.cpsc474.CribbageCard;
import edu.yale.cs.cpsc474.CribbageHand;
import edu.yale.cs.cpsc474.PeggingHistory;
import edu.yale.cs.cpsc474.CribbageGame;

public class MyPolicy implements CribbagePolicy
{
    // the default implementation delegates decisions to an existing policy
    private CompoundPolicy policy;
    
    public MyPolicy(CribbageGame game)
    {
	policy = new CompoundPolicy(new GreedyThrower(game), new GreedyPegger());
    }

    @Override
    public CribbageHand[] keep(CribbageHand cards, int[] scores, boolean amDealer)
    {
	return policy.keep(cards, scores, amDealer);
    }

    @Override
    public CribbageCard peg(CribbageHand cards, PeggingHistory hist, int[] scores, boolean amDealer)
    {
	return policy.peg(cards, hist, scores, amDealer);
    }
}

File: code/cpp/my_policy.cpp
#include "my_policy.h"

#include <vector>

#include <climits>

#include "game.h"
#include "hand.h"
#include "policy.h"
#include "greedy.h"

namespace cpsc474
{
  MyPolicy::MyPolicy(const CribbageGame& game)
    : keepPolicy(new GreedyThrower(game)),
      pegPolicy(new GreedyPegger(game)),
      policy(new CompoundPolicy(*keepPolicy, *pegPolicy))
  {
  }

  MyPolicy::~MyPolicy()
  {
    delete policy;
    delete keepPolicy;
    delete pegPolicy;
  }
  
  std::vector<CribbageHand *> MyPolicy::keep(const CribbageHand& cards, const std::vector<int>& scores, bool amDealer)
  {
    return policy->keep(cards, scores, amDealer);
  }

  const CribbageCard *MyPolicy::peg(const CribbageHand& cards, const PeggingHistory& hist, const std::vector<int>& scores, bool amDealer)
  {
    return policy->peg(cards, hist, scores, amDealer);
  }

}
